### PR TITLE
NODE-98, feat: bitcoin registration pool

### PR DESCRIPTION
**Pull request by dnjscksdn98**
## Description

This pull request describes new changes required for Bitcoin address registrations.

- Implement `pallet_btc_registration_pool`
    - purpose: verifies user registration requests and manages registered addresses (Bitcoin and Bifrost both)
- Implement `precompile_btc_registration_pool`
    - purpose: creates `pallet_btc_registration_pool` accessible to the Ethereum interface

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Something else (simple changes that are not related to existing functionality or others)

# Checklist

- [x] I have selected the correct base branch.
- [x] I have performed a self-review of my own code.
- [ ] I have made corresponding changes to the documentation.
- [x] I have made new test codes regarding to my changes.
- [x] I have no personal secrets or credentials described on my changes.
- [x] I have run `cargo-clippy`  and linted my code.
- [x] My changes generate no new warnings.
- [x] My changes passed the existing test codes.
- [x] My changes are able to compile.

### CODE REVIEW

**Review by SunoukMun — APPROVED** [review]

File: pallets/btc-registration-pool/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+
+use sp_core::{ConstU32, RuntimeDebug};
+use sp_runtime::{BoundedBTreeMap, BoundedVec, DispatchError};
+use sp_std::vec::Vec;
+
+use parity_scale_codec::{Decode, Encode, MaxEncodedLen};
+use scale_info::TypeInfo;
+
+/// The maximum length of a valid Bitcoin address in characters (~62 alphanumeric characters).
+pub const ADDRESS_MAX_LENGTH: u32 = 62;
```
Comment (alstjd0921) [inline pallets/btc-registration-pool/src/lib.rs:17]:
```suggestion
pub const ADDRESS_MAX_LENGTH: u32 = 90;
```

File: pallets/btc-registration-pool/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+
+/// The maximum amount of accounts a multi-sig account can consist.
+pub const MULTI_SIG_MAX_ACCOUNTS: u32 = 16;
```
Comment (alstjd0921) [inline pallets/btc-registration-pool/src/lib.rs:20]:
```suggestion
pub const MULTI_SIG_MAX_ACCOUNTS: u32 = 20;
```

정책상 16으로 정해진게 아니면, n은 20까지 지정될 수 있긴합니다.

File: pallets/btc-registration-pool/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+
+/// The Bitcoin address type (length bounded).
+pub type BoundedBitcoinAddress = BoundedVec<u8, ConstU32<ADDRESS_MAX_LENGTH>>;
```
Comment (alstjd0921) [inline pallets/btc-registration-pool/src/lib.rs:23]: u8? char?

File: pallets/btc-registration-pool/src/pallet/impls.rs [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+		Ok(pub_keys)
+	}
+
+	/// Build the script for p2wsh address creation.
+	fn build_redeem_script(pub_keys: Vec<PublicKey>) -> Builder {
+		let mut redeem_script =
+			Builder::new().push_opcode(Opcode::from(<RequiredM<T>>::get().saturating_add(80))); // m
+
+		for key in pub_keys.iter() {
+			redeem_script = redeem_script.push_key(&key);
```
Comment (alstjd0921) [inline pallets/btc-registration-pool/src/pallet/impls.rs:33]:
```suggestion
			redeem_script = redeem_script.push_key(key);
```

Comment (alstjd0921) [reply pallets/btc-registration-pool/src/pallet/impls.rs:33]: key: &PublicKey

File: pallets/btc-registration-pool/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+			AddressState::Generated(a) => a == address,
+		}
+	}
+
+	pub fn is_generation_ready<T: Config>(&self) -> bool {
+		<RequiredN<T>>::get() as usize == self.pub_keys.len()
+	}
+
+	pub fn is_key_submitted(&self, pub_key: &Public) -> bool {
+		self.pub_keys.values().cloned().collect::<Vec<Public>>().contains(pub_key)
```
Comment (alstjd0921) [inline pallets/btc-registration-pool/src/lib.rs:86]:
```suggestion
		self.pub_keys.values().into_iter().any(|x| x == pub_key)
```

File: pallets/btc-registration-pool/src/pallet/mod.rs [hunk 6]
```diff
@@ -0,0 +1,375 @@
...
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		#[pallet::call_index(0)]
+		#[pallet::weight(<T as Config>::WeightInfo::set_vault_config())]
+		pub fn set_vault_config(origin: OriginFor<T>, m: u8, n: u8) -> DispatchResultWithPostInfo {
+			T::SetOrigin::ensure_origin(origin)?;
+
+			ensure!(n >= 1 && n <= 16, Error::<T>::OutOfRange);
```
Comment (alstjd0921) [inline pallets/btc-registration-pool/src/pallet/mod.rs:169]:
```suggestion
			ensure!(n >= 1 && n <= 20, Error::<T>::OutOfRange);
```

File: primitives/multi-sig/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+			AddressState::Generated(a) => a == address,
+		}
+	}
+
+	pub fn is_key_generation_ready(&self) -> bool {
+		self.n as usize == self.pub_keys.len()
+	}
+
+	pub fn is_key_submitted(&self, pub_key: &Public) -> bool {
+		self.pub_keys.values().cloned().collect::<Vec<Public>>().contains(pub_key)
```
Comment (alstjd0921) [inline primitives/multi-sig/src/lib.rs:88]:
```suggestion
		self.pub_keys.values().into_iter().any(|x| x == pub_key)
```

File: primitives/multi-sig/src/lib.rs [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+use sp_std::vec::Vec;
+
+use parity_scale_codec::{Decode, Encode, MaxEncodedLen};
+use scale_info::TypeInfo;
+
+/// The maximum amount of accounts a multi-sig account can consist.
+pub const MULTI_SIG_MAX_ACCOUNTS: u32 = 16;
+
+/// The maximum length of a valid Bitcoin address in characters (~64 alphanumeric characters).
+pub const ADDRESS_MAX_LENGTH: u32 = 64;
```
Comment (alstjd0921) [inline primitives/multi-sig/src/lib.rs:21]:
```suggestion
/// The maximum length of a valid Bitcoin address in characters (~90 alphanumeric characters).
pub const ADDRESS_MAX_LENGTH: u32 = 90;
```

File: precompiles/btc-socket-queue/src/lib.rs [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+	#[precompile::view]
+	fn unsigned_psbts(handle: &mut impl PrecompileHandle) -> EvmResult<Vec<UnboundedBytes>> {
+		handle.record_cost(RuntimeHelper::<Runtime>::db_read_gas_cost())?;
+
+		let mut psbts = vec![];
+		pallet_btc_socket_queue::PendingRequests::<Runtime>::iter().for_each(|(_, request)| {
+			psbts.push(UnboundedBytes::from(request.unsigned_psbt));
+		});
+
+		Ok(psbts)
```
Comment (alstjd0921) [inline precompiles/btc-socket-queue/src/lib.rs:41]:
```suggestion
		let psbts: Vec<UnboundedBytes> =
			pallet_btc_socket_queue::PendingRequests::<Runtime>::iter()
				.map(|(_, request)| UnboundedBytes::from(request.unsigned_psbt))
				.collect();
		Ok(psbts)
```

**Review by alstjd0921 — APPROVED** [review]

File: pallets/btc-registration-pool/src/pallet/impls.rs [hunk 10]
```diff
@@ -2,6 +2,7 @@ use bp_multi_sig::{
 	traits::{MultiSigManager, PoolManager},
 	Address, AddressState, Network, UnboundedBytes,
 };
+use miniscript::Descriptor;
```
Comment (dnjscksdn98) [inline pallets/btc-registration-pool/src/pallet/impls.rs:5]: `bp-multi-sig`로 옮겨서 Cargo.toml에 제거해도 될것 같습니다

Comment (alstjd0921) [reply pallets/btc-registration-pool/src/pallet/impls.rs]: 완료

**Pull request merged**